src/qibo/tomography/gate_set_tomography.py

```diff
@@ -0,0 +1,302 @@
+from typing import Union
+
+import numpy as np
+from networkx import connectivity
+from sympy import S
+
+from qibo import Circuit, gates, symbols
+from qibo.backends import GlobalBackend, _check_backend
+from qibo.config import raise_error
+from qibo.hamiltonians import SymbolicHamiltonian
+from qibo.transpiler.optimizer import Preprocessing
+from qibo.transpiler.pipeline import Passes
+from qibo.transpiler.placer import Random
+from qibo.transpiler.router import Sabre
+from qibo.transpiler.unroller import NativeGates, Unroller
+
+SUPPORTED_NQUBITS = [1, 2]
+"""Supported nqubits for GST."""
+
+
+def _check_nqubits(nqubits):
+    if nqubits not in SUPPORTED_NQUBITS:
+        raise_error(
+            ValueError,
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2.",
+        )
+
+
+@cache
+def _gates(nqubits):
+    """Gates implementing all the GST state preparations.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.gates.abstract.Gate`): list of the gates used to prepare the possible states.
+    """
+
+    return list(
+        product(
+            [(gates.I,), (gates.X,), (gates.H,), (gates.H, gates.S)], repeat=nqubits
+        )
+    )
+
+
+@cache
+def _measurements(nqubits):
+    """Measurement gates implementing all the GST measurement bases.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.gates.abstract.Gate`): list of the gates implementing the possible measurement bases.
+    """
+
+    return list(product([gates.Z, gates.X, gates.Y, gates.Z], repeat=nqubits))
+
+
+@cache
+def _observables(nqubits):
+    """All the observables measured in GST.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.symbols.Symbol`): list of the possible observables measured.
+    """
+
+    return list(product([symbols.I, symbols.Z, symbols.Z, symbols.Z], repeat=nqubits))
+
+
+@cache
+def _get_observable(j, nqubits):
+    """Returns the j-th observable. Here,
+
+    .. math::
+        j \\in \\{0, 1, 2, 3\\}^{\\otimes n} \\equiv \\{ I, X, Y, Z\\}^{\\otimes n}.
+
+    Args:
+        j (int): The index of the measurement basis.
+        nqubits (int): Number of qubits.
+    Returns:
+        list(:class:`qibo.hamiltonians.SymbolicHamiltonian`): Symbolic hamiltonian of the observable.
+    """
+
+    if j == 0:
+        _check_nqubits(nqubits)
+    observables = _observables(nqubits)[j]
+    observable = S(1)
+    for q, obs in enumerate(observables):
+        if obs is not symbols.I:
+            observable *= obs(q)
+    return SymbolicHamiltonian(observable, nqubits=nqubits)
+
+
+@cache
+def _prepare_state(k, nqubits):
+    """Prepares the :math:`k`-th state for an :math:`n`-qubits (`nqubits`) circuit, where
+
+    .. math::
+        k \\in \\{0, 1, 2, 3\\}^{\\otimes n} \\equiv \\{ 0\\rangle\\langle0|, |1\\rangle\\langle1|,
```

Comment:
Same index issue here

Reply:
Same as above. k are labels. So k \in {0, 1, 2, 3} refers to preparing the states as {|0>, |1>, |+>, |y+>}. If we do the \otimes 2 for 2 qubit system, then we have k \in {0, 1, 2, 3, 4, ..., 15} refers to preparing the states {|0>|0>, |0>|1>, |0>|+>, |0>|y+>, |1>|0>, ..., |y+>|y+>}.

Reply:
Same reply here